comp110-worksheet-6/OxoBoard.cs

```diff
@@ -10,38 +10,136 @@ public enum Mark { None, O, X };
 
 	public class OxoBoard
 	{
-		// Constructor. Perform any necessary data initialisation here.
-		// Uncomment the optional parameters if attempting the stretch goal -- keep the default values to avoid breaking unit tests.
-		public OxoBoard(/* int width = 3, int height = 3, int inARow = 3 */)
+        private List<List<Mark>> board;
+
+        // Constructor. Perform any necessary data initialisation here.
+        // Uncomment the optional parameters if attempting the stretch goal -- keep the default values to avoid breaking unit tests.
+        public OxoBoard(int width = 3, int height = 3, int inARow = 3)
 		{
-			throw new NotImplementedException("TODO: implement this function and then remove this exception");
+            board = new List<List<Mark>>();
+            for (int i = 0; i < height; i++)
+            {
+                List<Mark> sublist = new List<Mark>();
+                for (int j = 0; j < width; j++)
+                {
+                    sublist.Add(Mark.None);
+                }
+                board.Add(sublist);
+            }
 		}
 
 		// Return the contents of the specified square.
 		public Mark GetSquare(int x, int y)
 		{
-			throw new NotImplementedException("TODO: implement this function and then remove this exception");
+            try
+            {
+                return board[x][y];
+            }
+            catch (System.ArgumentOutOfRangeException)
+            {
+                return Mark.X;
+            }
 		}
 
 		// If the specified square is currently empty, fill it with mark and return true.
 		// If the square is not empty, leave it as-is and return False.
 		public bool SetSquare(int x, int y, Mark mark)
 		{
-			throw new NotImplementedException("TODO: implement this function and then remove this exception");
+            if (GetSquare(x, y) == Mark.None)
+            {
+                board[x][y] = mark;
+                return true;
+            }
+            else
+            {
+                return false;
+            }
+            
 		}
 
 		// If there are still empty squares on the board, return false.
 		// If there are no empty squares, return true.
 		public bool IsBoardFull()
 		{
-			throw new NotImplementedException("TODO: implement this function and then remove this exception");
+            bool hasSpace = true;
+            foreach (List<Mark> row in board)
+            {
+                foreach (Mark mark in row)
+                {
+                    if (mark == Mark.None)
+                    {
+                        hasSpace = false;
+                    }
+                }
+            }
+            return hasSpace;
 		}
 
+        private Mark checkLine(List<Mark> line)
+        {
+            Mark winner = line[0];
+            Mark firstMark = line[0];
+            foreach (Mark mark in line)
+            {
+                if ((firstMark == Mark.None) || (mark != firstMark))
+                {
+                    winner = Mark.None;
```

Comment:
Similar to previous comment, you could `return Mark.None` here (for obvious reasons, ending a loop early is good for efficiency)